File: app/src/main/java/br/com/algorit/mangaalert/retrofit/MangaService.java
package br.com.algorit.mangaalert.retrofit;

import android.os.AsyncTask;

import java.lang.ref.WeakReference;
import java.util.List;

import br.com.algorit.mangaalert.roomdatabase.model.Manga;
import br.com.algorit.mangaalert.roomdatabase.model.Novel;
import br.com.algorit.mangaalert.retrofit.callback.CallbackWithReturn;
import retrofit2.Call;

public class MangaService {

    private final MangaRepository mangaRepository;

    public MangaService() {
        mangaRepository = new MangaRetrofit().getMangaRepository();
    }

    public void findAll(ResponseCallback<List<Manga>> callback) {
        WeakReference<MangaRepository> weakMangaRepository = new WeakReference<>(mangaRepository);
        new FindMangasAsyncTask(weakMangaRepository).execute(callback);
    }

    public void findAllNovel(ResponseCallback<List<Novel>> callback) {
        WeakReference<MangaRepository> weakMangaRepository = new WeakReference<>(mangaRepository);
        new FindNovelsAsyncTask(weakMangaRepository).execute(callback);
    }

    public interface ResponseCallback<T> {
        void success(T response);

        void fail(String erro);
    }

    private static class FindMangasAsyncTask extends AsyncTask<ResponseCallback<List<Manga>>, Void, Void> {

        private final WeakReference<MangaRepository> weakMangaRepository;

        FindMangasAsyncTask(WeakReference<MangaRepository> weakMangaRepository) {
            this.weakMangaRepository = weakMangaRepository;
        }

        @Override
        protected Void doInBackground(ResponseCallback<List<Manga>>... callbacks) {
            Call<List<Manga>> call = weakMangaRepository.get().findAll();
            call.enqueue(new CallbackWithReturn<>(new CallbackWithReturn.ResponseCallback<List<Manga>>() {
                @Override
                public void success(List<Manga> response) {
                    callbacks[0].success(response);
                }

                @Override
                public void fail(String fail) {
                    callbacks[0].fail(fail);
                }
            }));
            return null;
        }
    }

    private static class FindNovelsAsyncTask extends AsyncTask<ResponseCallback<List<Novel>>, Void, Void> {

        private final WeakReference<MangaRepository> weakMangaRepository;

        FindNovelsAsyncTask(WeakReference<MangaRepository> weakMangaRepository) {
            this.weakMangaRepository = weakMangaRepository;
        }

        @Override
        protected Void doInBackground(ResponseCallback<List<Novel>>... callbacks) {
            Call<List<Novel>> call = weakMangaRepository.get().findAllNovel();
            call.enqueue(new CallbackWithReturn<>(new CallbackWithReturn.ResponseCallback<List<Novel>>() {
                @Override
                public void success(List<Novel> response) {
                    callbacks[0].success(response);
                }

                @Override
                public void fail(String fail) {
                    callbacks[0].fail(fail);
                }
            }));
            return null;
        }
    }

}

File: app/src/main/java/br/com/algorit/mangaalert/retrofit/MangaService.kt
package br.com.algorit.mangaalert.retrofit

import android.os.AsyncTask
import br.com.algorit.mangaalert.retrofit.callback.CallbackWithReturn
import br.com.algorit.mangaalert.roomdatabase.model.Manga
import br.com.algorit.mangaalert.roomdatabase.model.Novel
import java.lang.ref.WeakReference

class MangaService {
    private val mangaRepository: MangaRepository = MangaRetrofit().mangaRepository

    fun findAll(callback: ResponseCallback<List<Manga>>) {
        val weakMangaRepository = WeakReference(mangaRepository)
        FindMangasAsyncTask(weakMangaRepository).execute(callback)
    }

    fun findAllNovel(callback: ResponseCallback<List<Novel>>) {
        val weakMangaRepository = WeakReference(mangaRepository)
        FindNovelsAsyncTask(weakMangaRepository).execute(callback)
    }

    interface ResponseCallback<T> {
        fun success(response: T)
        fun fail(erro: String)
    }

    private class FindMangasAsyncTask(private val weakMangaRepository: WeakReference<MangaRepository>) :
        AsyncTask<ResponseCallback<List<Manga>>, Void, Void>() {
        override fun doInBackground(vararg params: ResponseCallback<List<Manga>>): Void? {
            val call = weakMangaRepository.get()!!.findAll()
            call.enqueue(CallbackWithReturn(object :
                CallbackWithReturn.ResponseCallback<List<Manga>> {
                override fun success(response: List<Manga>) {
                    params[0].success(response)
                }

                override fun fail(fail: String) {
                    params[0].fail(fail)
                }
            }))
            return null
        }
    }

    private class FindNovelsAsyncTask(private val weakMangaRepository: WeakReference<MangaRepository>) :
        AsyncTask<ResponseCallback<List<Novel>>, Void, Void>() {
        override fun doInBackground(vararg params: ResponseCallback<List<Novel>>): Void? {
            val call = weakMangaRepository.get()!!.findAllNovel()
            call.enqueue(CallbackWithReturn(object :
                CallbackWithReturn.ResponseCallback<List<Novel>> {
                override fun success(response: List<Novel>) {
                    params[0].success(response)
                }

                override fun fail(fail: String) {
                    params[0].fail(fail)
                }
            }))
            return null
        }
    }
}